contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1426 @@
+        override
+        onlyOwnerOrOperator(_operator)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        (, uint96 tRemainder) = tToNu(_amount);
+        _amount -= tRemainder;
+        require(
+            _amount > 0 &&
+                _amount + getMinStaked(_operator, StakingProvider.NU) <=
+                operator.nuInTStake,
+            "Can't unstake specified amount of tokens"
+        );
+        operator.nuInTStake -= _amount;
+
+        emit Unstaked(_operator, _amount);
+    }
+
+    /// @notice Sets cached legacy stake amount to 0, sets the liquid T stake
+    ///         amount to 0 and withdraws all liquid T from the stake to the
+    ///         owner. Reverts if there is at least one non-zero authorization.
+    ///         Can be called only by the delegation owner and operator.
+    function unstakeAll(address _operator)
+        external
+        override
+        onlyOwnerOrOperator(_operator)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        require(
+            operator.authorizedApplications.length == 0,
+            "At least one application is still authorized"
+        );
+        require(
+            operator.tStake == 0 ||
+                minTStakeAmount == 0 ||
+                operator.startTStakingTimestamp + MIN_STAKE_TIME <=
+                /* solhint-disable-next-line not-rely-on-time */
+                block.timestamp,
+            "Unstaking is possible only after 24 hours"
+        );
+
+        emit Unstaked(
+            _operator,
+            operator.tStake + operator.keepInTStake + operator.nuInTStake
+        );
+        uint96 amount = operator.tStake;
+        operator.tStake = 0;
+        operator.keepInTStake = 0;
+        operator.nuInTStake = 0;
+        if (amount > 0) {
+            token.safeTransfer(operator.owner, amount);
+        }
+    }
+
+    //
+    //
+    // Keeping information in sync
+    //
+    //
+
+    /// @notice Notifies about the discrepancy between legacy KEEP active stake
+    ///         and the amount cached in T staking contract. Slashes the operator
+    ///         in case the amount cached is higher than the actual active stake
+    ///         amount in KEEP staking contract. Needs to update authorizations
+    ///         of all affected applications and execute an involuntary
+    ///         authorization decrease on all affected applications. Can be called
+    ///         by anyone, notifier receives a reward.
+    function notifyKeepStakeDiscrepancy(address _operator) external override {
+        OperatorInfo storage operator = operators[_operator];
+        require(operator.keepInTStake > 0, "Nothing to slash");
+
+        (uint256 keepStakeAmount, , uint256 undelegatedAt) = keepStakingContract
+            .getDelegationInfo(_operator);
+        (uint96 tAmount, ) = keepToT(keepStakeAmount);
+
+        require(
+            operator.keepInTStake > tAmount || undelegatedAt != 0,
```

Comment:
should this also have a check for undelegatedAt being in the past?

Reply:
I see that no difference when `undelegatedAt` in future or in past. Because in both cases it's violation and we will set staked Keep to zero